### PR TITLE
feat: Limit number of spoke reserves via spoke configurator

### PR DESCRIPTION
**Pull request by CheyenneAtapour**
* Enforces a limit on the spoke configurator such that when a reserve is added to a spoke, the operation reverts if that additional reserve would exceed the allowed limit on the spoke
* The owner of the spoke configurator can configure the limit of reserves per each spoke

The motivation for this is that if a reserve has an arbitrary number of spokes (and we do not enforce limits on numbers of collateral on the user level), a user can enable so many collaterals that it causes any operations on their position to fail because they exceed gas limits. The main cause of this is the sorting algorithm involved in calculating a user's risk premium. Indeed, analysis shows that 166 collaterals is the limit where calculating risk premium succeeds, but 167 fails. The other operations such as supplying, borrowing, or liquidating will still succeed and are only in the realm of 17m gas, vs the limit of 45m. The analysis was performed for the worst case of the quicksort algorithm, in which the collaterals are already in sorted order (of collateral risk).

Such analysis should be performed on each spoke implementation to determine the appropriate reserve limit for each spoke

closes https://github.com/aave/aave-v4/issues/734

### CODE REVIEW

File: src/spoke/SpokeConfigurator.sol [hunk 1]
```diff
@@ -14,6 +14,8 @@ import {ISpokeConfigurator} from 'src/spoke/interfaces/ISpokeConfigurator.sol';
 contract SpokeConfigurator is Ownable2Step, ISpokeConfigurator {
   using SafeCast for uint256;
 
+  mapping(address => uint256) internal _reserveLimit;
```
Comment (yan-man) [inline src/spoke/SpokeConfigurator.sol]: if it's internal perhaps we should also create a getter for this?

File: src/spoke/SpokeConfigurator.sol [hunk 2]
```diff
@@ -68,6 +70,11 @@ contract SpokeConfigurator is Ownable2Step, ISpokeConfigurator {
     ISpoke(spoke).updateLiquidationConfig(liquidationConfig);
   }
 
+  /// @inheritdoc ISpokeConfigurator
+  function updateReserveLimit(address spoke, uint256 limit) external onlyOwner {
+    _reserveLimit[spoke] = limit;
```
Comment (yan-man) [inline src/spoke/SpokeConfigurator.sol]: maybe also emit an event here?

should we check limit > 0? hm ig limit=0 is valid 

and shouldnt be an issue, but also then we're ok with setting limit < current reserveCount too

File: src/spoke/interfaces/ISpokeConfigurator.sol [hunk 3]
```diff
@@ -8,6 +8,11 @@ import {ISpoke} from 'src/spoke/interfaces/ISpoke.sol';
 /// @author Aave Labs
 /// @notice Interface for the SpokeConfigurator.
 interface ISpokeConfigurator {
+  /// @dev Thrown when the reserve limit for a spoke is reached.
+  /// @param spoke The address of the spoke.
+  /// @param limit The reserve limit.
```
Comment (yan-man) [inline src/spoke/interfaces/ISpokeConfigurator.sol]: maybe can clarify here that limit is inclusive

should we rename to cap vs limit for consistency as well?

File: tests/unit/SpokeConfigurator.t.sol [hunk 4]
```diff
@@ -180,7 +180,40 @@ contract SpokeConfiguratorTest is SpokeBase {
     });
   }
 
+  function test_addReserve_revertsWith_ReserveLimitReached() public {
+    vm.prank(SPOKE_CONFIGURATOR_ADMIN);
+    spokeConfigurator.updateReserveLimit(spokeAddr, 0);
```
Comment (yan-man) [inline tests/unit/SpokeConfigurator.t.sol]: can you add another test scenario for a non-zero value

File: src/spoke/interfaces/ISpokeConfigurator.sol (hunk 3 above)
Comment (CheyenneAtapour) [reply src/spoke/interfaces/ISpokeConfigurator.sol]: We don't specify inclusive for the other caps.

I like limit because cap can be confusing with supply or draw cap

File: src/spoke/SpokeConfigurator.sol (hunk 2 above)
Comment (CheyenneAtapour) [reply src/spoke/SpokeConfigurator.sol]: Emitted event

File: tests/unit/SpokeConfigurator.t.sol (hunk 4 above)
Comment (CheyenneAtapour) [reply tests/unit/SpokeConfigurator.t.sol]: Updated this test to use potentially non-zero values (using random uint)

File: src/spoke/SpokeConfigurator.sol (hunk 1 above)
Comment (CheyenneAtapour) [reply src/spoke/SpokeConfigurator.sol]: Added

Comment (miguelmtzinf) [inline src/spoke/SpokeConfigurator.sol]:
```suggestion
  mapping(address spoke => uint256) internal _maxNumberReserves;
```

Comment (CheyenneAtapour) [reply src/spoke/SpokeConfigurator.sol]: Updated, but @miguelmtzinf did you want the getters and everything to change names as well? I kinda like the name as is for the external functions etc.

Reason being `maxNumberReserves` is not proper grammar, because it should be `maxNumberOfReserves`, or we could just do `maxReserves`?

File: src/spoke/SpokeConfigurator.sol [hunk 5]
```diff
@@ -68,6 +70,12 @@ contract SpokeConfigurator is Ownable2Step, ISpokeConfigurator {
     ISpoke(spoke).updateLiquidationConfig(liquidationConfig);
   }
 
+  /// @inheritdoc ISpokeConfigurator
+  function updateReserveLimit(address spoke, uint256 limit) external onlyOwner {
```
Comment (yan-man) [inline src/spoke/SpokeConfigurator.sol]: maybe these names should be updated to reflect new var name, ie updateMaxNumberReserves or something

File: src/spoke/interfaces/ISpokeConfigurator.sol (hunk 3 above)
Comment (yan-man) [reply src/spoke/interfaces/ISpokeConfigurator.sol]: maybe also update the natspec to reflect new var naming too

**Comment by yan-man** [comment]
btw @CheyenneAtapour also closes this one: https://github.com/aave/aave-v4/issues/734

File: src/spoke/SpokeConfigurator.sol (hunk 5 above)
Comment (CheyenneAtapour) [reply src/spoke/SpokeConfigurator.sol]: Done, I went with just `maxReserves`

File: src/spoke/interfaces/ISpokeConfigurator.sol (hunk 3 above)
Comment (CheyenneAtapour) [reply src/spoke/interfaces/ISpokeConfigurator.sol]: Done, I went with just `maxReserves`

**Comment by CheyenneAtapour** [comment]
I kinda like `limit` better personally, but nbd

File: src/spoke/interfaces/ISpokeConfigurator.sol [hunk 6]
```diff
@@ -38,6 +48,11 @@ interface ISpokeConfigurator {
     ISpoke.LiquidationConfig calldata liquidationConfig
   ) external;
 
+  /// @notice Updates the maximum number of reserves allowed to exist on a spoke.
+  /// @param spoke The address of the spoke.
```
Comment (miguelmtzinf) [inline src/spoke/interfaces/ISpokeConfigurator.sol:53]:
```suggestion
  /// @dev It allows to set the maximum below the amount of reserves that currently exist.
  /// @param spoke The address of the spoke.
```

File: src/spoke/interfaces/ISpokeConfigurator.sol [hunk 7]
```diff
@@ -8,6 +8,16 @@ import {ISpoke} from 'src/spoke/interfaces/ISpoke.sol';
...
 interface ISpokeConfigurator {
+  /// @notice Emitted when the maximum allowed number of reserves for a spoke is updated.
+  /// @param spoke The address of the spoke.
+  /// @param maxReserves The new maximum number of reserves.
+  event MaxReservesUpdated(address spoke, uint256 maxReserves);
+
+  /// @dev Thrown upon adding a reserve when the maximum allowed number of reserves is already reached.
+  /// @param spoke The address of the spoke.
+  /// @param maxReserves The maximum allowed number of reserves.
+  error MaxReservesReached(address spoke, uint256 maxReserves);
```
Comment (miguelmtzinf) [inline src/spoke/interfaces/ISpokeConfigurator.sol]:
```suggestion
  error MaximumReservesReached(address spoke, uint256 maxReserves);
```

Comment (CheyenneAtapour) [reply src/spoke/interfaces/ISpokeConfigurator.sol]: How about the name of the event? Is that fine?

File: src/spoke/interfaces/ISpokeConfigurator.sol (hunk 6 above)
Comment (CheyenneAtapour) [reply src/spoke/interfaces/ISpokeConfigurator.sol:53]: Changed, I worded it slightly differently

File: src/spoke/interfaces/ISpokeConfigurator.sol [hunk 8]
```diff
@@ -8,6 +8,16 @@ import {ISpoke} from 'src/spoke/interfaces/ISpoke.sol';
 /// @author Aave Labs
 /// @notice Interface for the SpokeConfigurator.
 interface ISpokeConfigurator {
+  /// @notice Emitted when the maximum allowed number of reserves for a spoke is updated.
+  /// @param spoke The address of the spoke.
+  /// @param maxReserves The new maximum number of reserves.
+  event MaxReservesUpdated(address spoke, uint256 maxReserves);
```
Comment (miguelmtzinf) [inline src/spoke/interfaces/ISpokeConfigurator.sol]:
```suggestion
  event UpdateMaxReserves(address spoke, uint256 maxReserves);
```
we are doing present tense verb + noum

Comment (CheyenneAtapour) [reply src/spoke/interfaces/ISpokeConfigurator.sol]: Done
